### DISCUSSION TITLE
Using electrodb with a dependency injection pattern

### DISCUSSION THREAD

**Discussion by bush** [tywalch/electrodb#337] [General]
Lets say we have a class and we want to design it with testing in mind.  I'd like to pass an Entity as a dependency:

```
// foo.ts
class Foo{
  private entity: Entity<any, any, any, any>
  constructor(entity: Entity<any, any, any, any>) {
    this.entity = entity;
  }
}

const entity = new Entity(...);
const foo = new Foo(entity);
```

Is there a better way of doing this?  If someone wants to do this they would throw away the strong type checking ...  I'm open to suggestions here ...

**Comment by tywalch** [comment]
Can you go into a bit more detail about your use-case? I think even with a type of `Entity<any, any, any, any>` you're still likely to run into cases where you'll need `// @ts-ignore` to use `private entity`

**Reply by bush** [reply]
Hi Tyler  - yes! and thanks for responding so quickly. In general I like to design my classes in a way that I can test them independent of the application I'm using them in.  So in the electrodb documentation the entity is imported into the file it's being used in. But if you want to make your class more testable a common way to do that is to pass in the dependencies to the object you are creating.  So in the case above I'll pass the entity into the constructor of the class when I instantiate an instance of it.  When I write a test case for my class I will create a test Entity object and pass it to the constructor of the class that is using it and execute series of test cases as such.  If the class that I am using imports the entity rather than, say,  taking an Entity as a parameter to the constructor its more difficult to test.  The problem I am having however is outlined above.  There is no straight forward pattern to specifying the type of the entity in your class.

**Reply by tywalch** [reply]
I love to inject dependencies too, so I understand that part of the need 👍 

The part I'm stuck on is that it appears your class needs to accept a sort of "generic" entity. Since the ElectroDB interface is dynamic based on the specifics of the entity, there is not much on a that could be done on a "generic" entity. Would you be able to share a snippet of how you'd like to work with this entity in a perfect world? Specifically how your class methods might interact with a generic entity?

**Reply by bush** [reply]
Thats the part I am stuck on as well ... I honestly am not sure how to do this cleanly without using the any escape hatch

I currently have something like this:

```
class ElectroDBTodoStorage implements TodoStorageProvider {

  // I have no easy way to describe the type
  private todo: NimkeeDBTodoEntity;

  constructor(todo: NimkeeDBTodoEntity) {
    this.todo = todo; 
  }

  async create(user: NimkeeDBTodoUser) {
    // where this.todo is the Entity
    this.todo.create(user);
  }
}
```
If import is the only way to do this thats cool I just wanted to make sure there wasn't something I was missing.  I'll take any other suggestions if you have any.  Much appreciated!

**Reply by bush** [reply]
Essentially what I am trying to do is hide your library behind an interface.  I want to write adapters that implement my interface for other databases as well.

**Reply by tywalch** [reply]
> Essentially what I am trying to do is hide your library behind an interface. I want to write adapters that implement my interface for other databases as well.

For sure, this is definitely ideal!

Given everything you said, I think your facade class needs to know actual entity it is accepting, not just a generic entity. Fortunately this is simple, here's an example using the playground: 

https://electrodb.fun/?#code/JYWwDg9gTgLgBAbzgUQHY2DAngGjgZQFMoA3YAY0LgF84AzKCEOAIkIBtDyZGATAIxYBuAFAjyEVAGd4MAIb9OcALyssEAK5QA+vMWFtqOSELCxE6fAAqEXhBVxUhAO4p0mLAAoRcRD99wILYcAFx+AQGE7thhLDC2ECw4-hEkxFLAkrEAjEkpAVLEZJSx8lIA1nJgYCz51MkRcjA8wPwaMIRSYQj5vsC83b0B2GCEsTJQwKgA5nkRvvVDqBAdg-PDWKPjLTNz63BQhACOGsCHA3A8GoRDi+vkh02EvACCMGv7I2OsqBog-MQ9uteIQ6HINOx3nBPABKFQAPjgABEngA6ZbOWENfYAehxcHIclQy3gAMCtmAdGAzzgcjoHSgBMeHV4Q18j14AHlUOwsGErjd1nd5howLwnq8oT19pdNt8WL9-oDses8XBnE1yAALejQAlaonTTqXexErC05qTNodNnqzVa2IAKiB8zVhXg5C0h3Ql1AnXk4HVWqicFF4pZtvdYVhCORaIxWNtILBEKhMeUiJRHXREExMJV8w53N5-Kg1wLCzqBamIIAHp0PhEyuUuuF9mByo39lSOBcWB2XfcmJAMqs4ABtAC6tuF6wqXfWPfYfYqg-mEnAEFH33HLH6LGnMuotwLx4W2KQeiUx5hohEXzgNjsaAw2AcXwgdEfCTvD6fEAASQ6ZhVGlOB+jCCYpmmURfBJb4oJmURj3vOVvzsAAFRgyBBABhZlCE5MAMEkKQHDAiC4EQmD-HgyCdholCpgZMFKHQiB8HiKA5CNLCIBw4g2weQgnk8CBiMyaQwn-PiBPwkSOiIkjpBhMI+JAYBCgAHn-IDCBAeFkLEB9kE4bhGCRAAhf9OOgHjCIk0jyP8KJXz5diXw8IzxHYOQpDI0yuB4CArJsrj7PA8BOBMdAyLCuzeOw-pBJ6fw1QAuADTSRx7BEqRzQ1c14jgEEpAeVoqBgYNZVGfwwEmEgnhNOxpISTzsDvXwLAmDRuGgMTHKklAzOC0KEls7ijSUySpDhMDfCqzTUXiOwHHE5SpFRVyvLgfwUN8PysFQcgmQUwgBo21rMKSvCCOm0i5vybr4CQeC8H6Gg1sGqRYICZ6Dk6VMHDkDVMEuLUlpWiBUWE0TXpWQh3t4GgYVRaYIFhX7fHyQ4YC0VAAakVNUXDORfuPFD-qhiaItUJxXEC8yQus8bwqNTwEG2mJ2OoW9jNZhLCBhgiOcogBybIACYAGYxbwOi4DF0AwGi1z6AUrRCDF3mhCAA

**Reply by bush** [reply]
Hey thats not bad!  I think I can work with that!  I noticed you didn't inject the client into the Todo Entity.  Does it automatically create one?  I guess thats the tricky part if I wanted to pass in the database client as a parameter I wouldn't be able to do that in this example.

**Reply by tywalch** [reply]
I do pass it in, on line 89!

**Reply by bush** [reply]
Oh shoot I missed that.  My humble appologies.

**Reply by tywalch** [reply]
No worries, I hope this helps!
